src/model/openthaigpt_pretraining_model/models/llama_hf/model.py

```diff
@@ -0,0 +1,508 @@
+    ]:
+        """
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
+            attention_mask (`torch.FloatTensor`, *optional*): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+            output_attentions (`bool`, *optional*):
+                Whether or not to return the attentions tensors of all attention layers. See `attentions` under
+                returned tensors for more detail.
+            use_cache (`bool`, *optional*):
+                If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding
+                (see `past_key_values`).
+            past_key_value (`Tuple(torch.FloatTensor)`, *optional*): cached past key and value projection states
+        """  # noqa
+        residual = hidden_states
+        hidden_states = self.input_layernorm(hidden_states)
+
+        # Self Attention
+        if gradient_checkpointing and self.training:
+
+            def create_custom_forward(module):
+                def custom_forward(*inputs):
+                    return module(
+                        *inputs,
+                        output_attentions=output_attentions,
+                        use_cache=use_cache,
+                    )
+
+                return custom_forward
+
+            (
+                hidden_states,
+                self_attn_weights,
+                present_key_value,
+            ) = torch.utils.checkpoint(
+                create_custom_forward(self.self_attn),
+                hidden_states,
+                attention_mask,
+                position_ids,
+                past_key_value,
+                output_attentions=output_attentions,
+                use_cache=use_cache,
+            )  # type: ignore
+        else:
+            hidden_states, self_attn_weights, present_key_value = self.self_attn(
+                hidden_states=hidden_states,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                past_key_value=past_key_value,
+                output_attentions=output_attentions,
+                use_cache=use_cache,
+            )
+
+        hidden_states = residual + hidden_states
+
+        # Fully Connected
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (self_attn_weights,)  # type: ignore
+
+        if use_cache:
+            outputs += (present_key_value,)  # type: ignore
+
+        return outputs  # type: ignore
+
+
+# create new Llama that call LlamaClass which edit LlamaDecoderLayerWithCheckpointing
+class LlamaNewCheckpoint(LlamaPreTrainedModel):
```

Comment:
Then you can just inherit class LlamaForCausalLMWithCheckpoint(LlamaForCausalLM) like this and just modify few functions